R/install.R

```diff
@@ -23,145 +66,31 @@
 #' @seealso [conda-tools], [virtualenv-tools]
 #'
 #' @export
-py_install <- function(
-  packages,
-  envname = NULL,
-  method = c("auto", "virtualenv", "conda"),
-  conda = "auto",
-  ...) {
+py_install <- function(packages,
+                       envname = Sys.getenv("RETICULATE_PYTHON_ENV", unset = "r-reticulate"),
+                       method = c("auto", "virtualenv", "conda"),
+                       conda = "auto",
+                       ...)
+{
+  # resolve 'auto' method
+  method <- match.arg(method)
+  if (method == "auto")
+    method <- py_install_method_detect(envname = envname, conda = conda)
 
   # validate method
   if (identical(method, "virtualenv") && is_windows()) {
     stop("Installing Python packages into a virtualenv is not supported on Windows",
          call. = FALSE)
   }
 
-  # when envname is NULL, use default-supplied environment
-  # (or r-reticulate environment otherwise for backwards compatibility)
-  if (is.null(envname))
-    envname <- Sys.getenv("RETICULATE_PYTHON_ENV", unset = "r-reticulate")
-
-  # find out which methods we can try
-  method_available <- function(name) any(method %in% c("auto", name))
-  virtualenv_available <- method_available("virtualenv")
-  conda_available <- method_available("conda")
-
-  # resolve and look for conda
-  conda <- tryCatch(conda_binary(conda), error = function(e) NULL)
-  have_conda <- conda_available && !is.null(conda)
-
-  # mac and linux
-  if (is_unix()) {
-
-    # check for explicit conda method
-    if (identical(method, "conda")) {
-
-      # validate that we have conda
-      if (!have_conda)
-        stop("Conda installation failed (no conda binary found)\n", call. = FALSE)
-
-      # do install
-      conda_install(envname, packages = packages, conda = conda, ...)
-
-    } else {
-
-      # find system python binary
-      pyver <- ""
-      python <- python_unix_binary("python")
-      if (is.null(python)) {
-        # try for python3 if we are on linux
-        if (is_linux()) {
-          python <- python_unix_binary("python3")
-          if (is.null(python))
-            stop("Unable to locate Python on this system.", call. = FALSE)
-          pyver <- "3"
-        }
-      }
-      # find other required tools
-      pip <- python_unix_binary(paste0("pip", pyver))
-      have_pip <- !is.null(pip)
-      virtualenv <- python_unix_binary("virtualenv")
-      have_virtualenv <- virtualenv_available && !is.null(virtualenv)
-
-      # if we don't have pip and virtualenv then try for conda if it's allowed
-      if ((!have_pip || !have_virtualenv) && have_conda) {
-
-        conda_install(envname, packages = packages, conda = conda, ...)
-
-      # otherwise this is either an "auto" installation w/o working conda
-      # or it's an explicit "virtualenv" installation
-      } else {
```

Comment:
My biggest worry about this change is that we are losing the printing of bootstrapping commands at the console. But perhaps these aren't necessary as we can assume they have pip / venv in most cases? In would be interesting to see what happens on a brand new OS X machine -- can the user get a working Python environment w/o help from us?

Reply:
Apple is actually planning to deprecate Python in the next release of macOS (https://developer.apple.com/documentation/macos_release_notes/macos_catalina_10_15_beta_3_release_notes) so we might want to consider suggesting to the user that they should install their own copy of Python.

I've had good luck with Homebrew Python and I think it's in general the way most people install Python + other software on macOS nowadays, so perhaps we should consider recommending that?

Reply:
I'll see if I can bring back installation instructions as well though (I agree they're useful to have for users)

Reply:
It looks like we do still provide recommendations on how pip / virtualenv / venv can be installed, but we defer it to environment creation time now:

https://github.com/rstudio/reticulate/blob/feature/py-install-revamp/R/virtualenv.R#L273-L295